miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+
+def mem(field):
+    """Generate a MemStruct subclass from a field. The field's value can
+    be accessed through self.value or self.deref_value if field is a Ptr.
+    """
+    if field in DYN_MEM_STRUCT_CACHE:
+        return DYN_MEM_STRUCT_CACHE[field]
+
+    fields = [("value", field)]
+    # Build a type to contain the field type
+    mem_type = type("Mem%r" % field, (MemStruct,), {'fields': fields})
+    DYN_MEM_STRUCT_CACHE[field] = mem_type
+    return mem_type
+
+
+# MemField classes
+
+class MemField(object):
+    """Base class to provide methods to set and get fields from virtual mem.
+
+    Subclasses can either override _pack and _unpack, or get and set if data
+    serialization requires more work (see Inline implementation for an example).
+    """
+
+    _self_type = None
+
+    def _pack(self, val):
+        """Serializes the python value @val to a raw str"""
+        raise NotImplementedError()
+
+    def _unpack(self, raw_str):
+        """Deserializes a raw str to an object representing the python value
+        of this field.
+        """
+        raise NotImplementedError()
+
+    def set(self, vm, addr, val):
+        """Set a VmMngr memory from a value.
+
+        Args:
+            vm: VmMngr instance
+            addr: the start adress in memory to set
+            val: the python value to serialize in @vm at @addr
+        """
+        raw = self._pack(val)
+        vm.set_mem(addr, raw)
+
+    def get(self, vm, addr):
+        """Get the python value of a field from a VmMngr memory at @addr."""
+        raw = vm.get_mem(addr, self.size())
+        return self._unpack(raw)
+
+    def _get_self_type(self):
+        return self._self_type
+
+    def _set_self_type(self, self_type):
+        """If this field refers to MemSelf, replace it with @self_type (a
+        MemStruct subclass) when using it. Generally not used outside the lib.
+        """
+        self._self_type = self_type
+
+    def size(self):
+        """Return the size in bytes of the serialized version of this field"""
+        raise NotImplementedError()
+
+    def __len__(self):
+        return self.size()
+
+    def __neq__(self, other):
+        return not self == other
+
+
+class Struct(MemField):
+    """Dumb struct.pack/unpack field. Mainly used to factorize code.
+
+    Value is a tuple corresponding to the struct @fmt passed to the constructor.
+    """
+
+    def __init__(self, fmt):
+        self._fmt = fmt
+
+    def _pack(self, fields):
+        return struct.pack(self._fmt, *fields)
+
+    def _unpack(self, raw_str):
+        return struct.unpack(self._fmt, raw_str)
+
+    def size(self):
+        return struct.calcsize(self._fmt)
+
+    def __repr__(self):
+        return "%s(%s)" % (self.__class__.__name__, self._fmt)
+
+    def __eq__(self, other):
+        return self.__class__ == other.__class__ and self._fmt == other._fmt
+
+    def __hash__(self):
+        return hash(hash(self.__class__) + hash(self._fmt))
```

Comment:
I think it is functionally equivalent, but your version is cleaner + shorter, I'll fix that.